### ISSUE TITLE
Rewrite rules ignore modalities

### ISSUE THREAD

**Issue by anuyts** [agda/agda#3839]
In the following code, we can prove `thm` using `refl` because of the rewrite rule, but the rewrite rule itself cannot be called in the same proof.

The example also works for irrelevance instead of shape-irrelevance, but then I expected people might argue that it is not a problem.

The problem also occurs in agda-parametric (where @jorisceulemans pointed it out to me) and likely also in agda-flat.

I expect that in many models of modal type systems, it will not be sound to take a definitional equality judgement and replace the modalities in its context with stronger modalities, which is exactly what this behaviour does.

```agda
{-# OPTIONS --experimental-irrelevance --rewriting #-}

module RewriteModal where

open import Data.Bool
open import Relation.Binary.PropositionalEquality

postulate
  f : ..(x : Bool) → Bool
  rw-f : {x : Bool} → f x ≡ true

{-# BUILTIN REWRITE _≡_ #-}
{-# REWRITE rw-f #-}

thm : ..{x : Bool} → f x ≡ true
thm {x} = refl

thm' : ..{x : Bool} → f x ≡ true
thm' {x} = rw-f {{!x!}}
```

**Comment by jespercockx** [comment]
Interesting. What would be the expected behaviour here? Should Agda just refuse this rewrite rule or should Agda try to actually do something smart?

**Comment by anuyts** [comment]
I would recommend simply refusing.
Many a headache can probably be prevented by explaining to the user what's going on when a term doesn't type check with modality-aware rewriting but does with modality-unaware rewriting...

**Comment by Saizan** [comment]
I think there are valid use cases for rewrite rules like this.

What needs to be done for soundness is checking that the terms the rule is instantiated with actually typecheck at the requested modality. I guess currently we do not have a cheap way to do that?

**Comment by anuyts** [comment]
I agree with @Saizan , I think I misunderstood @jespercockx in my last reaction...
